url2b4a2d655b_Yehonal_unisa_prog2_jdm_tgz-pJ8-com_jdeals_tests_TestUserBuyingJ8/src/com/jdeals/libs/JTextFieldFilter.java

```diff
@@ -0,0 +1,136 @@
+            = "abcdefghijklmnopqrstuvwxyz";
+
+    /**
+     * The Constant UPPERCASE.
+     */
+    public static final String UPPERCASE
+            = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+    /**
+     * The Constant ALPHA.
+     */
+    public static final String ALPHA
+            = LOWERCASE + UPPERCASE;
+
+    /**
+     * The Constant NUMERIC.
+     */
+    public static final String NUMERIC
+            = "0123456789";
+
+    /**
+     * The Constant FLOAT.
+     */
+    public static final String FLOAT
+            = NUMERIC + ".";
+
+    /**
+     * The Constant ALPHA_NUMERIC.
+     */
+    public static final String ALPHA_NUMERIC
+            = ALPHA + NUMERIC;
+
+    /**
+     * The Constant DATE.
+     */
+    public static final String DATE
+            = NUMERIC + "/";
+
+    /**
+     * The accepted chars.
+     */
+    protected String acceptedChars = null;
+
+    /**
+     * The negative accepted.
+     */
+    protected boolean negativeAccepted = false;
+
+    /**
+     * Instantiates a new j text field filter.
+     */
+    public JTextFieldFilter() {
+        this(ALPHA_NUMERIC);
+    }
+
+    /**
+     * Instantiates a new j text field filter.
+     *
+     * @param acceptedchars the acceptedchars
+     */
+    public JTextFieldFilter(String acceptedchars) {
+        acceptedChars = acceptedchars;
+    }
+
+    /**
+     * Sets the negative accepted.
+     *
+     * @param negativeaccepted the new negative accepted
+     */
+    public void setNegativeAccepted(boolean negativeaccepted) {
+        if (acceptedChars.equals(NUMERIC)
+                || acceptedChars.equals(FLOAT)
+                || acceptedChars.equals(ALPHA_NUMERIC)) {
+            negativeAccepted = negativeaccepted;
+            acceptedChars += "-";
+        }
+    }
+
+    /* (non-Javadoc)
+     * @see javax.swing.text.PlainDocument#insertString(int, java.lang.String, javax.swing.text.AttributeSet)
+     */
+    public void insertString(int offset, String str, AttributeSet attr)
+            throws BadLocationException {
+        if (str == null) {
+            return;
+        }
+
+        if (acceptedChars.equals(UPPERCASE)) {
+            str = str.toUpperCase();
+        } else if (acceptedChars.equals(LOWERCASE)) {
+            str = str.toLowerCase();
+        }
+
+        for (int i = 0; i < str.length(); i++) {
+            if (acceptedChars.indexOf(str.valueOf(str.charAt(i))) == -1) {
+                return;
+            }
+        }
+
+        if (acceptedChars.equals(FLOAT)
+                || (acceptedChars.equals(FLOAT + "-") && negativeAccepted)) {
+            if (str.indexOf(".") != -1) {
+                if (getText(0, getLength()).indexOf(".") != -1) {
+                    return;
+                }
+            }
+        }
+
+        if (negativeAccepted && str.indexOf("-") != -1) {
+            if (str.indexOf("-") != 0 || offset != 0) {
+                return;
+            }
+        }
+
+        super.insertString(offset, str, attr);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;attr&#39; (index: 2 | type: AttributeSet). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;offset&#39;, &#39;str&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).